### ISSUE TITLE
Dropout isn't deterministic between Intel and M1 MacBooks

### ISSUE THREAD

**Issue by PetrochukM** [pytorch/pytorch#84030]
### 🐛 Describe the bug

Hello, Folks!

I ran this code on two different laptops, and I got different results. They are both running the same PyTorch versions. 

```python
import random

import numpy as np
import torch

random.seed(123)
torch.manual_seed(123)
np.random.seed(123)
tensor = torch.rand(10)
print(tensor.sum())
print(torch.dropout(tensor, p=0.5, train=True).sum())
```

On my M1 MacBook, I get:
```
tensor(3.8430)
tensor(2.4471)
```

On my 2015 MacBook, I got:
```
tensor(3.8430)
tensor(1.8894)
```

### Versions

2015 Macbook Pro
```
Collecting environment information...
PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.0.1 (x86_64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.8.13 (default, May  8 2022, 17:52:27)  [Clang 13.1.6 (clang-1316.0.21.2)] (64-bit runtime)
Python platform: macOS-12.0.1-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.4
[pip3] pytorch-lightning==1.6.4
[pip3] pytorch-nlp==0.5.0
[pip3] torch==1.12.1
[pip3] torchaudio==0.11.0
[pip3] torchmetrics==0.9.1
[conda] Could not collect
```

2021 M1 Macbook
```
Collecting environment information...
PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.5 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.8.13 (default, Aug 11 2022, 15:30:37)  [Clang 13.1.6 (clang-1316.0.21.2.5)] (64-bit runtime)
Python platform: macOS-12.5-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.4
[pip3] pytorch-nlp==0.5.0
[pip3] torch==1.12.1
[pip3] torchaudio==0.11.0
[conda] Could not collect
```

**Comment by PetrochukM** [comment]
I'd like this to be reopened @gchanan. 

These results are very far off `tensor(1.8894)` and `tensor(2.4471)`. While I do expect results to be somewhat different, I don't expect them to be completely different.

This is also preventing our ability to test PyTorch. We rely on PyTorch being deterministic, somewhat, for testing.

This hasn't been an issue in the past, so from my POV, this is a regression.

**Comment by albanD** [comment]
SInce there are only 10 values, it is expected for the variance to be quite high.
If you increase the size of the Tensor to `10000` for example, you should see a much smaller difference.

**Comment by PetrochukM** [comment]
@albanD This variance is high enough that the correctness tests for our models have completely different values. This doesn't average out to a small difference, at the end of the day.

**Comment by albanD** [comment]
There are two things here:
- rng is not expected to be reproducible from one machine/dtype/pytorch version/backend to another. So we do expect to see different results here.
- Dropout will select some values and so it will give you the same variable as a Tensor twice as small. So these are expected discrepencies. You can use randn and mean() to get something more easily interpretable.

**Comment by PetrochukM** [comment]
Okay. I think there might be a bug because `torch.rand` is consistent and `torch.dropout` is not consistent. I understand if `rng` is not consistent from platform to platform; however, it seems like the same `rng` is being used, and yet `dropout` produces a different outcome.

**Comment by albanD** [comment]
You can always print the output and make sure that on average half of the values are zeroed out?

**Comment by PetrochukM** [comment]
@albanD Yes! There are plenty of workarounds. If it turns out, that this is intended behavior and we can't rely on PyTorch to be reproducible or deterministic at all. 

Anyways, I'd love to get back to the core issue!

**Comment by malfet** [comment]
Hmm, perhaps something is weird for Mac x86?
As I see the same results for 1.12.1 between my M1 and Colab notebook:
```
python -c "import torch;torch.manual_seed(123); x = torch.rand(10); y=torch.dropout(x, p=0.5, train=True); print(x,y, x.sum(), y.sum())"

tensor([0.2961, 0.5166, 0.2517, 0.6886, 0.0740, 0.8665, 0.1366, 0.1025, 0.1841,
        0.7264]) tensor([0.0000, 0.0000, 0.0000, 0.0000, 0.1479, 0.0000, 0.2732, 0.2050, 0.3681,
        1.4529]) tensor(3.8430) tensor(2.4471)
```
